src/db_reader.rs

```diff
@@ -0,0 +1,1099 @@
+use crate::bytes_range::BytesRange;
+use crate::config::{
+    CheckpointOptions, Clock, DbReaderOptions, ReadOptions, ScanOptions, SystemClock,
+};
+use crate::db_reader::ManifestPollerMsg::Shutdown;
+use crate::db_state::CoreDbState;
+use crate::db_stats::DbStats;
+use crate::error::SlateDBError;
+use crate::manifest::store::{ManifestStore, StoredManifest};
+use crate::manifest::Manifest;
+use crate::mem_table::{ImmutableMemtable, ImmutableWal, KVTable};
+use crate::reader::{ReadSnapshot, Reader};
+use crate::sst::SsTableFormat;
+use crate::sst_iter::SstIteratorOptions;
+use crate::stats::StatRegistry;
+use crate::tablestore::TableStore;
+use crate::utils::MonotonicClock;
+use crate::wal_replay::{WalReplayIterator, WalReplayOptions};
+use crate::{utils, Checkpoint, DbIterator};
+use bytes::Bytes;
+use log::{info, warn};
+use object_store::path::Path;
+use object_store::ObjectStore;
+use once_cell::sync::Lazy;
+use parking_lot::{Mutex, RwLock};
+use std::collections::VecDeque;
+use std::ops::{RangeBounds, Sub};
+use std::sync::Arc;
+use tokio::runtime::Handle;
+use tokio::select;
+use tokio::sync::mpsc::{UnboundedReceiver, UnboundedSender};
+use uuid::Uuid;
+
+/// Read-only interface for accessing a database from either
+/// the latest persistent state or from an arbitrary checkpoint.
+pub struct DbReader {
+    inner: Arc<DbReaderInner>,
+    manifest_poller: Option<ManifestPoller>,
+}
+
+struct DbReaderInner {
+    manifest_store: Arc<ManifestStore>,
+    table_store: Arc<TableStore>,
+    options: DbReaderOptions,
+    state: Arc<RwLock<Arc<CheckpointState>>>,
+    clock: Arc<dyn Clock + Sync + Send>,
+    user_checkpoint_id: Option<Uuid>,
+    reader: Reader,
+}
+
+struct ManifestPoller {
+    join_handle: Mutex<Option<tokio::task::JoinHandle<Result<(), SlateDBError>>>>,
+    thread_tx: UnboundedSender<ManifestPollerMsg>,
+}
+
+enum ManifestPollerMsg {
+    Shutdown,
+}
+
+#[derive(Clone)]
+struct CheckpointState {
+    checkpoint: Checkpoint,
+    manifest: Manifest,
+    imm_memtable: VecDeque<Arc<ImmutableMemtable>>,
+    last_wal_id: u64,
+}
+
+static EMPTY_TABLE: Lazy<Arc<KVTable>> = Lazy::new(|| Arc::new(KVTable::new()));
+static EMPTY_WAL: Lazy<VecDeque<Arc<ImmutableWal>>> = Lazy::new(VecDeque::new);
+
+impl ReadSnapshot for Arc<CheckpointState> {
```

Comment:
I think it would be more idiomatic/flexible to do:
- `impl ReadSnapshot for CheckpointState`
- have the reader accept `&dyn ReadSnapshot`
- change the callers into the reader pass a reference to the snapshot impl